documentation/content/main/start-here/username-password.qwik.md

```diff
@@ -0,0 +1,592 @@
+
+```tsx
+// src/routes/signup/index.tsx
+import { component$ } from "@builder.io/qwik";
+import {
+	Form,
+	Link,
+	routeAction$,
+	routeLoader$,
+	z,
+	zod$
+} from "@builder.io/qwik-city";
+import { auth } from "~/lib/lucia";
+import { Prisma } from "@prisma/client";
+import { LuciaError } from "lucia-auth";
+
+// create an action to handle the form submission
+export const useSignupAction = routeAction$(
+	async (values, event) => {
+		try {
+			const user = await auth.createUser({
+				primaryKey: {
+					providerId: "username",
+					providerUserId: values.username,
+					password: values.password
+				},
+				attributes: {
+					username: values.username
+				}
+			});
+			const session = await auth.createSession(user.userId);
+			const authRequest = auth.handleRequest(event);
+			authRequest.setSession(session);
+		} catch (error) {
+			if (
+				error instanceof Prisma.PrismaClientKnownRequestError &&
+				error.code === "P2002" &&
+				error.message?.includes("username")
+			) {
+				return event.fail(500, {
+					error: "Username already in use"
+				});
+			}
+			if (
+				error instanceof LuciaError &&
+				error.message === "AUTH_DUPLICATE_KEY_ID"
+			) {
+				return event.fail(500, {
+					error: "Username already in use"
+				});
+			}
+			// database connection error
+			console.error(error);
+			return event.fail(500, {
+				error: "Unknown error occurred"
+			});
+		}
+
+		// if all goes well, redirect to home page
+		throw event.redirect(302, "/");
+	},
+	zod$({
+		username: z.string().min(3).max(20),
+		password: z.string().min(6).max(20)
+	})
+);
+```
+
+#### Handle requests
+
+Calling [`handleRequest()`] will create a new [`AuthRequest`](/referencel/lucia-auth/authrequest) instance, which makes it easier to handle sessions and cookies. This can be initialized with [`IncomingMessage`](https://nodejs.org/api/http.html#class-httpincomingmessage) and [`OutgoingMessage`](https://nodejs.org/api/http.html#class-httpserverresponse).
+
+In this case, we don't need to validate the request, but we do need it for setting the session cookie with [`AuthRequest.setSession()`](/reference/lucia-auth/authrequest#setsession).
+
+```ts
+const authRequest = auth.handleRequest(event);
+```
+
+#### Set user passwords
+
+We don't store the password in the user, but in the key (`primaryKey`). Keys represent the relationship between a user and a auth method, in this case username/password. We'll set `"username"` as the provider id (authentication method) and the username as the provider user id (something unique to the user).
+
+```ts
+const user = await auth.createUser({
+	primaryKey: {
+		providerId: "username",
+		providerUserId: username,
+		password
+	}
+	// ...
+});
+```
+
+### Add the action to the form
+
+We'll add the action to the form
+
+```tsx
+export default component$(() => {
+```
+
+```diff
++  const signupAction = useSignupAction();
```

Comment:
I don't think the diff works well, can you remove the diffing an just combine all the separated code blocks?

Reply:
oh sure! give a second 👍

Reply:
done!